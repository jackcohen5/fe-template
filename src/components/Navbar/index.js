import { memo } from "react"
import PropTypes from "prop-types"
import { Link as RouterLink, useRouteMatch } from "react-router-dom"
import { useSigninCheck, useFirestoreDocData, useFirestore } from "reactfire"
import { doc } from "firebase/firestore"

import routes from "routes"
import Button from "components/Button"
import Link from "components/Link"
import { useLogout, USER_COLLECTION } from "flux/ducks/auth"

import { ButtonWrapper, Container, AvatarImg, Name } from "./Navbar.styles"

const LoggedInNavbarAction = ({ userId }) => {
    const logout = useLogout()
    const { status, data: { firstName = null, lastName = null } = {} } =
        useFirestoreDocData(doc(useFirestore(), USER_COLLECTION, userId))

    return (
        <>
            <AvatarImg src="/default-avatar.png" alt="Avatar" />
            {status === "success" ? (
                <Name>
                    {firstName} {lastName}
                </Name>
            ) : null}
            <Button onClick={logout} ariaLabel="Logout">
                Logout
            </Button>
        </>
    )
}

LoggedInNavbarAction.propTypes = {
    userId: PropTypes.string.isRequired,
}

const useNavActions = ({ signedIn, user }) => {
    const { path } = useRouteMatch()

    if (signedIn && user) {
        return <LoggedInNavbarAction userId={user.uid} />
    } else {
        return path === routes.LOGIN ? null : (
            <RouterLink to={routes.LOGIN}>
                <Button ariaLabel="Login">Login</Button>
            </RouterLink>
        )
    }
}

const Navbar = () => {
    const { status, data: { signedIn = false, user = null } = {} } =
        useSigninCheck()
    const navActions = useNavActions({ signedIn, user })

    return (
        <Container>
            <Link to={routes.HOME} label="FE Template" />
            {status === "loading" ? null : (
                <ButtonWrapper data-testid="nav-actions-wrapper">
                    {navActions}
                </ButtonWrapper>
            )}
        </Container>
    )
}

export { NavbarHeight } from "./Navbar.constants"
export default memo(Navbar)
