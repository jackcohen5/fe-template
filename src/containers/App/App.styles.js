import styled from 'styled-components'

export const AppContainer = styled.div`
    margin: 50px 10px;
    overflow-y: scroll;
`

export const AppButton = styled.button`
    background-color: blue;
`
