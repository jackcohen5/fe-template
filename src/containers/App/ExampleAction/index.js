import { memo } from 'react'
import PropTypes from 'prop-types'

import { ExampleContainer, MarginedButton } from './ExampleAction.styles'

const ExampleAction = ({
    buttonDisabled,
    buttonLabel,
    onClick,
    description,
}) => {
    return (
        <ExampleContainer>
            <MarginedButton
                ariaLabel={buttonLabel}
                disabled={buttonDisabled}
                onClick={onClick}
                type="button"
            >
                {buttonLabel}
            </MarginedButton>
            {description}
        </ExampleContainer>
    )
}

ExampleAction.propTypes = {
    buttonDisabled: PropTypes.bool,
    buttonLabel: PropTypes.string.isRequired,
    description: PropTypes.node,
    onClick: PropTypes.func.isRequired,
}

ExampleAction.defaultProps = {
    description: null,
    buttonDisabled: false,
}

export default memo(ExampleAction)
