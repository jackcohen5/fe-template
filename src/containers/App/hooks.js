import { useMemo } from "react"
import { useLocation } from "react-router-dom"

import Loader from "components/Loader"
import { Roles } from "flux/ducks/auth"
import routes from "routes"

export const useLinkParams = (role = Roles.ROLE_1) => {
    const { pathname } = useLocation()
    if (pathname !== routes.HOME) {
        return {
            to: routes.HOME,
            label: "Go to public app",
        }
    }
    return role === Roles.ROLE_1
        ? {
              to: routes.ROLE_1_ROUTE,
              label: "Go to role 1 app",
          }
        : {
              to: routes.ROLE_2_ROUTE,
              label: "Go to role 2 app",
          }
}

export const useExampleAction = ({ exampleAction, hasTriggeredExample }) => {
    const description = useMemo(
        () => (
            <div>
                Example has {!hasTriggeredExample && "NOT "}been triggered.
            </div>
        ),
        [hasTriggeredExample]
    )
    return { description, buttonDisabled: false, onClick: exampleAction }
}

export const useExampleApiAction = ({
    apiExampleAction,
    exampleApiIsLoading,
    exampleApiResult,
    hasTriggeredExampleApi,
}) => {
    const description = useMemo(
        () =>
            exampleApiIsLoading ? (
                <>
                    <Loader />
                    <div>Calling in 1 second...</div>
                </>
            ) : (
                <>
                    <div>
                        Example API call has {!hasTriggeredExampleApi && "NOT "}
                        been triggered.
                    </div>
                    {hasTriggeredExampleApi ? (
                        <div>{exampleApiResult}</div>
                    ) : null}
                </>
            ),
        [hasTriggeredExampleApi, exampleApiIsLoading, exampleApiResult]
    )
    return {
        description,
        buttonDisabled: exampleApiIsLoading,
        onClick: apiExampleAction,
    }
}
