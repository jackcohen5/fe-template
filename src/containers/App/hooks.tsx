import { useMemo } from "react"
import { useLocation } from "react-router-dom"

import Loader from "components/Loader"
import { Role } from "flux/ducks/auth"
import Route from "routes"

export const useLinkParams = (role = Role.ROLE_1) => {
    const { pathname } = useLocation()
    if (pathname !== Route.HOME) {
        return {
            to: Route.HOME,
            label: "Go to public app",
        }
    }
    return role === Role.ROLE_1
        ? {
              to: Route.ROLE_1_ROUTE,
              label: "Go to role 1 app",
          }
        : {
              to: Route.ROLE_2_ROUTE,
              label: "Go to role 2 app",
          }
}

export const useExampleAction = ({
    exampleAction,
    hasTriggeredExample,
}: {
    exampleAction: () => void
    hasTriggeredExample: boolean
}) => {
    const description = useMemo(
        () => (
            <div>
                Example has {!hasTriggeredExample && "NOT "}been triggered.
            </div>
        ),
        [hasTriggeredExample]
    )
    return { description, buttonDisabled: false, onClick: exampleAction }
}

export const useExampleApiAction = ({
    apiExampleAction,
    exampleApiIsLoading,
    exampleApiResult,
    hasTriggeredExampleApi,
}: {
    apiExampleAction: () => void
    exampleApiIsLoading: boolean
    exampleApiResult: string
    hasTriggeredExampleApi: boolean
}) => {
    const description = useMemo(
        () =>
            exampleApiIsLoading ? (
                <>
                    <Loader />
                    <div>Calling in 1 second...</div>
                </>
            ) : (
                <>
                    <div>
                        Example API call has {!hasTriggeredExampleApi && "NOT "}
                        been triggered.
                    </div>
                    {hasTriggeredExampleApi ? (
                        <div>{exampleApiResult}</div>
                    ) : null}
                </>
            ),
        [hasTriggeredExampleApi, exampleApiIsLoading, exampleApiResult]
    )
    return {
        description,
        buttonDisabled: exampleApiIsLoading,
        onClick: apiExampleAction,
    }
}
