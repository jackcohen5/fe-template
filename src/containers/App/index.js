import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

import Button from 'src/components/Button'
import Navbar from 'src/components/Navbar'
import routes from 'src/routes'

import {
    exampleAction,
    hasTriggeredExampleSelector,
} from 'src/flux/ducks/example'

import { AppContainer } from './App.styles'

export const UnwrappedApp = ({ hasTriggeredExample, exampleAction }) => (
    <>
        <Navbar title="nametbd" />
        <AppContainer>
            Congrats, you started up nametbd.
            <br />
            Example has {!hasTriggeredExample && 'NOT '}been triggered.
            <br />
            <Button onClick={() => exampleAction()} type="button">
                Trigger Example Action
            </Button>
            <br />
            <Link to={routes.PROTECTED_HOME}>Go to private App</Link>
        </AppContainer>
    </>
)

UnwrappedApp.propTypes = {
    exampleAction: PropTypes.func.isRequired,
    hasTriggeredExample: PropTypes.bool.isRequired,
}

const mapStateToProps = (state) => ({
    hasTriggeredExample: hasTriggeredExampleSelector(state),
})

const mapDispatchToProps = {
    exampleAction,
}

export default connect(mapStateToProps, mapDispatchToProps)(UnwrappedApp)
