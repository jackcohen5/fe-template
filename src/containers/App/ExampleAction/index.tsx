import { memo } from "react"

import { ExampleContainer, MarginedButton } from "./ExampleAction.styles"

type ExampleActionProps = {
    buttonDisabled?: boolean
    buttonLabel: string
    onClick: () => void
    description: React.ReactNode
}

const ExampleAction = ({
    buttonDisabled = false,
    buttonLabel,
    onClick,
    description,
}: ExampleActionProps) => {
    return (
        <ExampleContainer>
            <MarginedButton
                ariaLabel={buttonLabel}
                disabled={buttonDisabled}
                onClick={onClick}
                type="button"
            >
                {buttonLabel}
            </MarginedButton>
            {description}
        </ExampleContainer>
    )
}

export default memo(ExampleAction)
