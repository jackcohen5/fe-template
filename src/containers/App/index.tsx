import { connect } from "react-redux"

import Link from "components/Link"
import Navbar from "components/Navbar"
import {
    apiExampleAction,
    exampleApiIsLoadingSelector,
    exampleApiResultSelector,
    hasTriggeredExampleApiSelector,
} from "flux/ducks/apiExample"
import { exampleAction, hasTriggeredExampleSelector } from "flux/ducks/example"
import { AppState } from "flux/store"

import { AppContainer } from "./App.styles"
import ExampleAction from "./ExampleAction"
import { useExampleAction, useExampleApiAction, useLinkParams } from "./hooks"

type AppProps = {
    apiExampleAction: () => void
    exampleAction: () => void
    exampleApiIsLoading: boolean
    exampleApiResult: string
    hasTriggeredExample: boolean
    hasTriggeredExampleApi: boolean
}

export const UnwrappedApp = ({
    apiExampleAction,
    exampleAction,
    exampleApiIsLoading,
    exampleApiResult,
    hasTriggeredExample,
    hasTriggeredExampleApi,
}: AppProps) => {
    const exampleActionProps = useExampleAction({
        exampleAction,
        hasTriggeredExample,
    })
    const exampleApiActionProps = useExampleApiAction({
        apiExampleAction,
        exampleApiIsLoading,
        exampleApiResult,
        hasTriggeredExampleApi,
    })
    const linkParams = useLinkParams()
    return (
        <>
            <Navbar />
            <AppContainer>
                <div>Congrats, you started up the FE template.</div>
                <ExampleAction
                    buttonLabel="Trigger Example Action"
                    {...exampleActionProps}
                />
                <ExampleAction
                    buttonLabel="Trigger API Call"
                    {...exampleApiActionProps}
                />
                <Link {...linkParams} />
            </AppContainer>
        </>
    )
}

const mapStateToProps = (state: AppState) => ({
    exampleApiIsLoading: exampleApiIsLoadingSelector(state),
    exampleApiResult: exampleApiResultSelector(state),
    hasTriggeredExample: hasTriggeredExampleSelector(state),
    hasTriggeredExampleApi: hasTriggeredExampleApiSelector(state),
})

const mapDispatchToProps = {
    apiExampleAction,
    exampleAction,
}

export default connect(mapStateToProps, mapDispatchToProps)(UnwrappedApp)
