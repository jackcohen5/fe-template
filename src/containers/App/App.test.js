import React from 'react'
import { shallow } from 'enzyme'

import Navbar from 'src/components/Navbar'

import { UnwrappedApp as App } from '.'
import { AppButton, AppContainer } from './App.styles'

const defaultProps = {
    hasTriggeredExample: false,
    exampleAction: () => {},
}

const renderComponent = (props = {}, method = shallow) =>
    method(<App {...defaultProps} {...props} />)

describe('App', () => {
    it('passes title to navbar', () => {
        const wrapper = renderComponent({})
        expect(wrapper.find(Navbar).exists()).toBe(true)
        expect(wrapper.find(Navbar).prop('title')).toBe('nametbd')
    })

    it('untriggered example matches snapshot', () => {
        const wrapper = renderComponent({
            hasTriggeredExample: false,
        })
        expect(wrapper.find(AppContainer).text()).toMatchSnapshot()
    })

    it('triggered example matches snapshot', () => {
        const wrapper = renderComponent({
            hasTriggeredExample: true,
        })
        expect(wrapper.find(AppContainer).text()).toMatchSnapshot()
    })

    it('calls exampleAction on button click', () => {
        const exampleActionSpy = jest.fn()
        const wrapper = renderComponent({
            exampleAction: exampleActionSpy,
        })
        wrapper.find(AppButton).simulate('click')
        expect(exampleActionSpy).toHaveBeenCalled()
    })
})
