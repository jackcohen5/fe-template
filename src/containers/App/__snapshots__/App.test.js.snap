// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App triggered example matches snapshot 1`] = `"Congrats, you started up nametbd.Example has been triggered.Trigger Example ActionGo to private App"`;

exports[`App untriggered example matches snapshot 1`] = `"Congrats, you started up nametbd.Example has NOT been triggered.Trigger Example ActionGo to private App"`;
